### ISSUE TITLE
Content Node with Many Complex Archetype Rows Pegs CPU for Minutes

### ISSUE THREAD

**Issue by Nicholas-Westby** [imulus/Archetype#244]
Note that this issue is not the same as this other issue, which is related to publish time (and is more a server-side issue): https://github.com/imulus/Archetype/issues/238

I have an Archetype property on my content node with about 72 of these rows:

![row](https://cloud.githubusercontent.com/assets/5933222/5598852/e70b8b88-9271-11e4-8072-af8f5d6b8c51.png)

When I click that content node in Chrome, "Google Chrome Helper" stays at 100% for 1 minute and 35 seconds:
![cpu](https://cloud.githubusercontent.com/assets/5933222/5598865/05dd2a30-9272-11e4-91f2-fd63ae1b2cbe.png)

This leads to a pretty painful content entry experience (it was reported by somebody who was entering content, and I confirmed on my machine).

**Comment by Nicholas-Westby** [comment]
Forgot to mention, Umbraco 7.1.8 and Archetype 1.6.0. The website is running on IIS (rather than Visual Studio). The data types have "Deep Datatype Requests" disabled.

**Comment by Nicholas-Westby** [comment]
After a bit of profiling and console.logging, it appears this may be the culprit:

![function](https://cloud.githubusercontent.com/assets/5933222/5622593/6c1fa98c-94fa-11e4-896e-a0f0e2e5cb6f.png)

In my case, that gets hit 372 times (once for each of the 4 properties on my 93 rows). And according to the Chrome profiler, it seems this function is taking up the vast majority of the CPU time (90% or more).

I'm not sure I quite understand the defer/resolve functionality, but it seems like it's suppose to return an object that can later be inspected for a value that gets set asynchronously. If that's the case, I'm not sure why this object gets created here and this is never referenced outside of this scope. I wonder if another function of a similar name was intended... maybe debounce?

My (admittedly uninformed) suspicion is that $rootScope.$apply() is supposed to refresh something or other. Maybe that should only happen once rather than 372 times? I'm not really sure, so I leave it to your expertise to figure out.

**Comment by kgiszewski** [comment]
Here is the commit message for the code in question: https://github.com/imulus/Archetype/commit/840feec96789444674f0688777d88b24ff6efd95

Have you tried removing that block or reducing the time?

**Comment by Nicholas-Westby** [comment]
Yes, it only takes a few seconds when I remove that block of code, though given the nature of that commit I'm not sure if this has any unintended side-effects (sounds like the commit prevents an intermittent issue). Maybe if debounce were used it would improve the performance without removing the fix outlined in that commit.

**Comment by Nicholas-Westby** [comment]
I am not at all sure what I'm doing, but I attempted to create a fix for this. First, I added this near the top of the JavaScript file:

![fix-1](https://cloud.githubusercontent.com/assets/5933222/5666896/26f3bef0-9718-11e4-8a0a-4675a11fb45f.png)

Then, I replaced the code from the commit you mentioned with this code:

![fix-2](https://cloud.githubusercontent.com/assets/5933222/5666903/3110e174-9718-11e4-8ed2-7d01581a0769.png)

I have an array of "$q" objects because I'm not sure if there can be different instances or if it's always the same instance. To avoid extra work, I check if each $q is the same (if so, I avoid the defer/resolve). I assume there is only one $rootScope, so I don't worry about checking for multiple instances.

Basically, I use debounce to avoid calling the same function repeatedly. As I said, I don't know if there are any side-effects, so use this at your own risk.

**Comment by kgiszewski** [comment]
Closing this due to #253.  We can come back to this if you think it's a thing still.

**Comment by Nicholas-Westby** [comment]
Yes, it's still a thing (this is a JavaScript issue, whereas the other one was a C# issue). Like I mentioned in the original post, these are two different issues:
![comment](https://cloud.githubusercontent.com/assets/5933222/5921808/5cb36468-a5fb-11e4-8256-12f23b6f4d81.png)

To clarify, the other issue was a server-side performance issue. This issue is a client-side performance issue.

By the way, my fix outlined above works for some nodes, but I've seen some long delays (~1 minute) on other nodes (maybe they just have more Archetype rows).

**Comment by kgiszewski** [comment]
Ok no problem.  Thanks for you efforts so far.  I re-opened and we can tackle this later.

**Comment by DaveScreenMedia** [comment]
Hey Guys,

Was this ever fixed?

We are having a similar issue

**Comment by kgiszewski** [comment]
Likely yes, if I were you I'd try the latest version of Archetype as there was a performance boost for large Archetypes a couple versions ago. #301 

In fact I'm going to close this for now and we will reopen it if there are any further reported issues.

**Comment by Nicholas-Westby** [comment]
I updated Archetype not too long ago on the site shown in the screenshots in my original report, and load times decreased from minutes to about 3 seconds. So yes, this has been fixed :-)

**Comment by DaveScreenMedia** [comment]
Thanks for this guys – I will fix as appropriate
